pkg/manifest/locator.go

```diff
@@ -114,6 +116,19 @@ func (l Locations) MarshalJSON() ([]byte, error) {
 	return json.Marshal(j)
 }
 
+// HTML extensions for [Locations]
+
+func (l Locations) CSSSelector() string {
+	if v, ok := l.OtherLocations["cssSelector"]; ok {
+		if s, ok := v.(string); ok {
+			return s
+		}
+	}
+	return ""
+}
+
+// TODO partialCfi and domRange getters
```

Comment:
Thorium doesn't really use / need EPUB CFI. We generate them in our "locators" next to the data structure we actually rely on to make things work internally (DOM Range for precise locations, percentage progression otherwise ... oh, and by the way we still don't support positions list yet). We emit CFI "just because we can", and it's only the path to the element anyway, no character-level reference. To achieve feature-parity with our internal DOM Range (de)serialisation mechanism, we would need a properly-tested and conformant CFI implementation, which apparently doesn't exist :)

* https://github.com/fread-ink/epub-cfi-resolver
* https://github.com/vivliostyle/vivliostyle.js/blob/master/packages/core/src/vivliostyle/cfi.ts
* https://github.com/johnfactotum/foliate-js/blob/main/epubcfi.js
* https://github.com/futurepress/epub.js/blob/master/src/epubcfi.js -- https://github.com/futurepress/EpubCFI/blob/master/src/epubcfi.js
* https://github.com/readium/readium-cfi-js
* https://github.com/satorumurmur/bibi/blob/master/__src/bibi/extensions/epubcfi.js
* Colibrio? (https://api-docs.colibrio.com/web/3.9.0/modules/colibrio_core_selector_epubcfi.html)

Also see https://github.com/futurepress/epub.js/issues/1358

Reply:
This work is based on the kotlin-toolkit, which is presumably based on the documentation here: https://github.com/readium/architecture/blob/master/models/locators/extensions/html.md . `partialCfi` is just a string in kotlin, so it could theoretically omitted since you can access it if it exists via `OtherLocations`, but since `DomRange` is [implemented in the kotlin-toolkit](https://github.com/readium/kotlin-toolkit/blob/develop/readium/shared/src/main/java/org/readium/r2/shared/publication/html/DomRange.kt) I would want to implement it in the go-toolkit as well

Reply:
The extension helpers are implemented but never used. It might not be worth implementing it in the Go toolkit, unless you do want to generate partial CFI or DOM ranges at some point.